sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable_connection.rs

```diff
@@ -0,0 +1,1055 @@
+    },
+    models::AmqpValue,
+    producer::DEFAULT_EVENTHUBS_APPLICATION,
+    ErrorKind, EventHubsError,
+};
+use async_lock::{Mutex as AsyncMutex, OnceCell};
+use azure_core::{
+    credentials::TokenCredential, error::ErrorKind as AzureErrorKind, http::Url, Result, Uuid,
+};
+use azure_core_amqp::{
+    error::{AmqpErrorCondition, AmqpErrorKind},
+    AmqpMessage, AmqpReceiver, AmqpReceiverApis, AmqpReceiverOptions, AmqpSendOptions,
+    AmqpSendOutcome, AmqpSessionOptions, AmqpSource,
+};
+use azure_core_amqp::{
+    AmqpClaimsBasedSecurity, AmqpClaimsBasedSecurityApis, AmqpConnection, AmqpConnectionApis,
+    AmqpConnectionOptions, AmqpError, AmqpManagement, AmqpManagementApis, AmqpOrderedMap,
+    AmqpSender, AmqpSenderApis, AmqpSenderOptions, AmqpSession, AmqpSessionApis, AmqpSimpleValue,
+    AmqpSymbol, AmqpTarget,
+};
+use futures::{select, FutureExt};
+use std::{collections::HashMap, error::Error, sync::Arc, time::Duration};
+use tracing::{debug, trace, warn};
+
+/// The recoverable connection is responsible for managing the connection to the Event Hubs service.
+/// It also handles authorization and connection recovery.
+///
+/// * Notes
+///
+/// The way a client uses a RecoverableConnection is as follows:
+///   1. Create a new instance of the RecoverableConnection.
+///   2. Retrieve an interim object from the RecoverableConnection. Supported
+///      interim objects are:
+///
+///    - `AmqpManagement`: Used for management operations.
+///    - `AmqpSender`: Used for sending messages to the Event Hubs service.
+///    - `AmqpReceiver`: Used for receiving messages from the Event Hubs service.
+///    - `AmqpClaimsBasedSecurity`: Used for authorization operations (should not be used directly)
+/// * 3. Use the interim object to perform operations on the Event Hubs service.
+///
+/// Under the covers, the interim objects contain a reference back to the [`RecoverableConnection`],
+/// and enough information to recreate the underlying AMQP connection, session, management, cbs, or sender/receiver
+/// objects as needed.
+///
+/// The various interim objects implement the appropriate AMQP APIs, but wrap the underlying APIs with
+/// a retry loop [`retry_azure_operation`], so that the actual client does not have to worry about retrying or recovering operations.
+///
+/// There is a taxonomy of methods in this struct:
+///   - `ensure_*` methods: These methods are used to ensure that the underlying connection, session, management client, cbs client, sender, or receiver is created and available.
+///   - `get_*` methods: These methods are used to retrieve a wrapper around the underlying session, management client, cbs client, sender, or receiver.
+///   - `create_*` methods: These methods are used to create a new underlying connection, session, management client, cbs client, sender, or receiver.
+///
+/// In general, the `ensure_*` and `create_*` methods are private to the ` RecoverableConnection`
```

Comment:
```suggestion
/// In general, the `ensure_*` and `create_*` methods are private to the `RecoverableConnection`
```

Also, you don't always mark code constructs with backticks above. Should be consistent.